### PR TITLE
Adding Flash Attention 2 Support for GPT2

### LINKED ISSUE #26350: Community contribution: Adding Flash Attention 2 support for more architectures
### Feature request

Flash Attention 2 is a library that provides attention operation kernels for faster and more memory efficient inference and training: https://github.com/Dao-AILab/flash-attention

Screenshot 2023-09-22 at 17 49 18

Let's try to add Flash Attention 2 support for more architectures! Currently supported architectures are

- [x] Llama
- [x] Falcon

It would be great to add the support for more architectures such as

- [x] Bark
- [x] Bart
- [ ] BERT | @sorenmc 
- [x] DistilBERT
- [ ] GPT-2
- [ ] GPT-J
- [x] GPTBigCode (Starcoder) | @susnato 
- [x] GPT-neo
- [x] GPT-neo-x | @younesbelkada #26463
- [x] OPT | @susnato #26414 
- [x] Llava
- [x] VipLlava
- [x] mBART
- [x] Mistral
- [x] Mixtral
- [ ] MPT | @rajveer43 
- [ ] T5
- [ ] Persimmon | @jeromeku 
- [x] Phi
- [x] Whisper
- [x] Qwen2

... and many more

Adding this feature would require to follow the same protocol as in https://github.com/huggingface/transformers/pull/25598 
 . First create a new module inside the corresponding modeling file termed as `xxxFlashAttention` that inherits from `xxxAttention` and override the foward method to use the public methods from `flash-attn`. Make sure to have access to a GPU that supports Flash Attention 2. 

Given the slight challenge of the issue, labelling it as a good second issue!

If you are interested to take up the challenge, comment below with the architecture name you want to integrate and open a PR!

Once you open a PR, feel free to ping @LysandreJik @ArthurZucker @amyeroberts @younesbelkada @fxmarty @SunMarc @pacman100 for a review

### Motivation

Making LLMs more memory efficient and faster ! 

### Your contribution

Reviewing PRs and possibly adding the support for more models

### PR DESCRIPTION
**Pull request by EduardoPach**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/26350

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Hey, @younesbelkada added flash attention 2 support for `GPT2`. The only thing missing is the Expected speedups, could you share the code you used for the other models you added support to keep consistency?

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Wow thanks for the great work ! At a quick glance it seems you took care very well of the copy mechanism which is quite a challenge for GPT2 !
Please find the benchmarking script: https://gist.github.com/younesbelkada/02f35734da906cc0f2389ae4f665c58f I suggest to try it out for prefill only on large sequence length - let us know with @ArthurZucker @fxmarty how it goes

**Comment by EduardoPach** [comment]
> Wow thanks for the great work ! At a quick glance it seems you took care very well of the copy mechanism which is quite a challenge for GPT2 ! Please find the benchmarking script: https://gist.github.com/younesbelkada/02f35734da906cc0f2389ae4f665c58f I suggest to try it out for prefill only on large sequence length - let us know with @ArthurZucker @fxmarty how it goes

Hey, I don't have a GPU and I was renting in RunPod an RTX 3090 to work on this PR, is it a problem to use the 3090 to benchmark or should I switch to an A100 (which I believe it was the GPU used in the other benchmarks at least the ones I've seen)?

**Comment by EduardoPach** [comment]
@ArthurZucker I believe it should be ready for review

**Review by younesbelkada — APPROVED** [review]
Great work !

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (ArthurZucker) [inline src/transformers/models/gpt2/modeling_gpt2.py:419]:
```suggestion
        present = None
        if use_cache is True:
            present = (key, value)
```
a small nit

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (ArthurZucker) [inline src/transformers/models/gpt2/modeling_gpt2.py:917]:
```suggestion
        self._attn_implementation = config._attn_implementation 
```
once we'll have sdpa as well this shall be re-used

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (ArthurZucker) [inline src/transformers/models/gpt2/modeling_gpt2.py:1039]:
```suggestion
            if self. _attn_implementation == "flash_attention_2":
```

**Review by ArthurZucker — COMMENTED** [review]
LGTM but we need to add a test 😉

**Comment by younesbelkada** [comment]
@EduardoPach thanks again, what @ArthurZucker meant is an integration test similar as: https://github.com/huggingface/transformers/blob/63caa370e6c618dbe7d3fd4cbf545cc32eca1a15/tests/models/llama/test_modeling_llama.py#L400 for GPT2 only, would you be happy to work on that? 🙏

**Comment by EduardoPach** [comment]
> @EduardoPach thanks again, what @ArthurZucker meant is an integration test similar as: https://github.com/huggingface/transformers/blob/63caa370e6c618dbe7d3fd4cbf545cc32eca1a15/tests/models/llama/test_modeling_llama.py#L400 for GPT2 only, would you be happy to work on that? 🙏 

Yeah, I will add the test in the following hours

File: docs/source/en/model_doc/gpt2.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/gpt2.md:65]:
```suggestion
The `generate()` method can be used to generate text using GPT2 model.
```

File: docs/source/en/model_doc/gpt2.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/gpt2.md:88]:
```suggestion
Flash Attention 2 is a faster, optimized version of the attention scores computation which relies on `cuda` kernels.
```

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (ArthurZucker) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:575]:
```suggestion
DECISION_TRANSFORMER_GPT2_ATTENTION_CLASSES = {
```

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (ArthurZucker) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:577]: Where is `DecisionTransformerGPT2FlashAttention2`

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (ArthurZucker) [inline tests/models/gpt2/test_modeling_gpt2.py:885]: Let's explicitly write `output_native` 😉

**Review by ArthurZucker — COMMENTED** [review]
Almost good, left a few nits

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (EduardoPach) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py]: https://github.com/huggingface/transformers/pull/29226/commits/74fb9bda09067c320d3fc28845aa897cd6e8fe09

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (EduardoPach) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py:578]: Haven't added it there, but added it now here https://github.com/huggingface/transformers/pull/29226/commits/74fb9bda09067c320d3fc28845aa897cd6e8fe09. However, `DecisionTransformer` does not support flash attention yet just I had to do these modifications to make sure nothing would break with the `Copy from` statements.

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (EduardoPach) [reply tests/models/gpt2/test_modeling_gpt2.py:885]: Not sure I get this one, the eager attention output is already called `output_native`

Comment (ArthurZucker) [reply tests/models/gpt2/test_modeling_gpt2.py:885]: What I mean is to just add:
```
output_native = ["THis is the expected native output", "And this as well"]` 
```
this helps us in the long run

**Review by ArthurZucker — APPROVED** [review]
LGMT on final nit for the test to have explicit values

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (EduardoPach) [reply tests/models/gpt2/test_modeling_gpt2.py:885]: Done here https://github.com/huggingface/transformers/pull/29226/commits/cafc0f12e8906f3a5234007d9b71fea86deae4a1

**Review by younesbelkada — APPROVED** [review]
Thanks for your great work @EduardoPach ! Can you merge this branch with upstream main? after that we can merge i think

**Comment by EduardoPach** [comment]
> Thanks for your great work @EduardoPach ! Can you merge this branch with upstream main? after that we can merge i think

Done! 😄

**Review by younesbelkada — APPROVED** [review]
Thanks again ! We just merged some fixes on main - could you rebase again 🙏 then we should finally merge :D sorry for all the iterations !

**Comment by EduardoPach** [comment]
> Thanks again ! We just merged some fixes on main - could you rebase again 🙏 then we should finally merge :D sorry for all the iterations !

No worries! Done

**Comment by younesbelkada** [comment]
Thanks ! Hmm I can't see the rebase commit on the history, perhaps can you try again ?

**Comment by EduardoPach** [comment]
> Thanks ! Hmm I can't see the rebase commit on the history, perhaps can you try again ?

I've done `git fetch upstream && git merge upstream/main` here https://github.com/huggingface/transformers/pull/29226/commits/dcde56c66907cdcd74d011fc7b1919446f8240c3

File: README_de.md
Comment (amyeroberts) [inline README_de.md:1]: There shouldn't be readme changes here. Can you make sure to rebase on main to include the mode recent changes?

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (amyeroberts) [inline src/transformers/models/gpt2/modeling_gpt2.py:1036]: Library convention is no asserts in modeling files 

```suggestion
```

File: docs/source/en/perf_infer_gpu_one.md
Comment (amyeroberts) [inline docs/source/en/perf_infer_gpu_one.md:1]: decision transformers should appear here too

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this and making our models go brrr 🔥 

Just a few small comments. The diffs in the READMEs will need to be resolved before we can merge

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (EduardoPach) [reply src/transformers/models/gpt2/modeling_gpt2.py:1036]: There are a bunch of `assert`s in the modeling file of `GPT2` which I suppose are legacy

File: docs/source/en/perf_infer_gpu_one.md
Comment (EduardoPach) [reply docs/source/en/perf_infer_gpu_one.md:1]: https://github.com/huggingface/transformers/pull/29226/commits/c36c60ae9256e17e2407895704a0e89ef7110503

File: docs/source/en/model_doc/gpt2.md
Comment (amyeroberts) [inline docs/source/en/model_doc/gpt2.md:63]: We should have the equivalent added for decision transformer too

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this! 

Only thing left to do is make sure decision transformer has the updated documentation and tests

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (EduardoPach) [reply tests/models/gpt2/test_modeling_gpt2.py:876]: Doesn't the test in GPT2 already cover Decision Transformer? Since, basically the usage of Flash Attention in Decision Transformer happens exactly due to GPT2Model being embedded in its architecture

Comment (amyeroberts) [reply tests/models/gpt2/test_modeling_gpt2.py:876]: Both model's should be tested. This makes sure that if anything changes upstream they remain correct, for example, inputs preparation in `DecisionTransformerModel`

Comment (EduardoPach) [reply tests/models/gpt2/test_modeling_gpt2.py:876]: While adding the test for `DecisionTransformer` I realized that the model has two distinct `xxxPreTrainedModel`s and that adding support for `flash_attention_2` would be a bit more complicated, therefore I believe it would be better to have a specific PR to add support.

File: docs/source/en/model_doc/gpt2.md
Comment (EduardoPach) [reply docs/source/en/model_doc/gpt2.md:63]: See message above

**Comment by EduardoPach** [comment]
> Thank you for your hard work! Many of us are excited about the GPT-2 model supporting flash attention. May I ask when the PR is expected to be merged?

Hey, I believe if @amyeroberts agrees with my latest message it should get merged right away 🤞

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (amyeroberts) [reply tests/models/gpt2/test_modeling_gpt2.py:876]: In this case, flash attention shouldn't be added at all for the model. You can use `#Ignore copy` on the init so the previous attention class' method is used

Comment (EduardoPach) [reply tests/models/gpt2/test_modeling_gpt2.py:875]: @amyeroberts https://github.com/huggingface/transformers/pull/29226/commits/94c2fe8859786fb1360584be352e06e9b4ff8ade

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (amyeroberts) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:554]: Here I would use # ignore copy - the model shouldn't have FA2 logic

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating - a few final places to tidy up.

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (EduardoPach) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py:552]: Addressed

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (EduardoPach) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py]: Addressed

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this for GPT2 and iterating on a solution!

**Comment by NieShenRuc** [comment]
It seems everything is okay. May I kindly request to merge this PR? I am really looking forward to speeding up my GPT-2. If my request has added to your workload, I apologize for any inconvenience.

**Comment by EduardoPach** [comment]
> It seems everything is okay. May I kindly request to merge this PR? I am really looking forward to speeding up my GPT-2. If my request has added to your workload, I apologize for any inconvenience.

c.c. @amyeroberts

**Pull request merged**
